File: use-cases/easy-ship/code/java/src/main/java/lambda/RetrieveOrderHandler.java
package lambda;

import java.util.*;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.Context;
import com.google.gson.Gson;
import io.swagger.client.api.OrdersV0Api;
import io.swagger.client.model.orders.EasyShipShipmentStatus;
import io.swagger.client.model.orders.GetOrderItemsResponse;
import io.swagger.client.model.orders.GetOrderResponse;
import io.swagger.client.model.orders.OrderItem;
import lambda.utils.*;

import static lambda.utils.Constants.REFRESH_TOKEN_KEY_NAME;
import static lambda.utils.Constants.REGION_CODE_KEY_NAME;
import static org.apache.commons.lang3.ObjectUtils.isEmpty;

public class RetrieveOrderHandler implements RequestHandler<StateMachineInput, StateMachineInput> {

    @Override
    public StateMachineInput handleRequest(StateMachineInput input, Context context) {
        LambdaLogger logger = context.getLogger();
        logger.log("ProcessNotification Lambda input: " + new Gson().toJson(input));

        try {
            // Validate input
            ValidateInput.validateRetrieveOrderInput(input);
            // Get Order V0 API instance
            OrdersV0Api OrdersApi = ApiUtils.getOrdersV0Api(input);

            // API calls to retrieve order and order items
            GetOrderResponse orderResponse = OrdersApi.getOrder(input.getAmazonOrderId());
            logger.log("Order Response : " + new Gson().toJson(orderResponse));

            GetOrderItemsResponse orderItemsResponse = OrdersApi.getOrderItems(input.getAmazonOrderId(), null);
            logger.log("Order Items Response : " + new Gson().toJson(orderItemsResponse));

            if (!EasyShipShipmentStatus.PENDINGSCHEDULE.equals(orderResponse.getPayload().getEasyShipShipmentStatus())) {
                    throw new IllegalArgumentException(
                        String.format("Amazon Order Id : %s is not EasyShip order", input.getAmazonOrderId()));
            }
            EasyShipOrder easyShipOrder = new EasyShipOrder();
            easyShipOrder.setOrderItems(getOrderItemList(orderItemsResponse));
            input.setEasyShipOrder(easyShipOrder);

            return input;
        } catch (Exception e) {
            throw new RuntimeException("Calling Orders API failed", e);
        }
    }


    private List<EasyShipOrderItem> getOrderItemList(GetOrderItemsResponse orderItems) {
        List<EasyShipOrderItem> itemList = new ArrayList<>();

        for (OrderItem orderItem : orderItems.getPayload().getOrderItems()) {
            EasyShipOrderItem item = new EasyShipOrderItem();
            item.setOrderItemId(orderItem.getOrderItemId());
            item.setSku(orderItem.getSellerSKU());
            item.setQuantity(orderItem.getQuantityOrdered());
            // Some region will need SerialNumber
            if (!isEmpty(orderItem.isSerialNumberRequired()) && orderItem.isSerialNumberRequired()) {
                item.setOrderItemSerialNumbers(orderItem.getSerialNumbers());
            }
            itemList.add(item);
        }

        return itemList;
    }
}

File: use-cases/easy-ship/code/php/lambda/RetrieveOrderHandler.php
<?php

namespace Lambda;

use GuzzleHttp\Exception\RequestException;
use InvalidArgumentException;
use Lambda\Utils\ApiUtils;
use Lambda\Utils\Model\EasyShipOrderItem;
use Lambda\Utils\LambdaContext;
use Lambda\Utils\Model\EasyShipOrder;
use Lambda\Utils\Model\StateMachineInput;
use Lambda\Utils\StateMachineInputConverter;
use SpApi\Model\orders\v0\EasyShipShipmentStatus;
use SpApi\Model\orders\v0\GetOrderItemsResponse;

class RetrieveOrderHandler
{
    public function handleRequest(array $event, LambdaContext $context): StateMachineInput
    {
        $input = StateMachineInputConverter::convertFromArray($event);
        $logger = $context->getLogger();
        $logger->info('RetrieveOrder Lambda input: ' . json_encode($event, JSON_PRETTY_PRINT));


        try {
            // Validate input
            $this->validateRetrieveOrderInput($input);

            // Get Orders V0 API instance
            $ordersApi = ApiUtils::getOrdersApi($input);

            // API call to retrieve order
            $orderResponse = $ordersApi->getOrder($input->getAmazonOrderId());

            // API call to retrieve order items
            $orderItemsResponse = $ordersApi->getOrderItems($input->getAmazonOrderId(), null);

            // Validate EasyShip order
            if (
                $orderResponse->getPayload()->getEasyShipShipmentStatus() !== EasyShipShipmentStatus::PENDING_SCHEDULE
            ) {
                throw new \InvalidArgumentException(sprintf(
                    'Amazon Order Id: %s is not an EasyShip order',
                    $input->getAmazonOrderId()
                ));
            }

            // Process order items and set to EasyShipOrder
            $input->setEasyShipOrder(
                new EasyShipOrder($this->getOrderItemList($orderItemsResponse)));

            return $input;
        } catch (RequestException $e) {
            $logger->error('RequestException: ' . $e->getMessage());
            throw new \RuntimeException('Calling Orders API failed', 0, $e);
        } catch (\Exception $e) {
            $logger->error('Exception: ' . $e->getMessage());
            throw new \RuntimeException('An unexpected error occurred', 0, $e);
        }
    }

    /**
     * @throws \Exception
     */
    private function getOrderItemList(GetOrderItemsResponse $orderItemsResponse): array
    {
        $itemList = [];
        foreach ($orderItemsResponse->getPayload()->getOrderItems() as $orderItem) {
            $item = new EasyShipOrderItem(
                $orderItem->getOrderItemId(),
                $orderItem->getSellerSku(),
                $orderItem->getQuantityOrdered(),
            );
            // Add SerialNumbers if required
            if ($orderItem->getSerialNumbers() !== null && $orderItem->getSerialNumberRequired()) {
                $item->setOrderItemSerialNumbers($orderItem->getSerialNumbers());
            }
            $itemList[] = $item;
        }
        return $itemList;
    }

    /**
     * @throws \Exception
     */
    private function validateRetrieveOrderInput(StateMachineInput $input): void
    {
        if ($input->getApiCredentials() === null) {
            throw new InvalidArgumentException("API credentials cannot be null");
        }
        if (empty($input->getAmazonOrderId())) {
            throw new InvalidArgumentException("Amazon Order Id cannot be null or empty");
        }
        if (empty($input->getMarketplaceId())) {
            throw new InvalidArgumentException("Marketplace Id cannot be null or empty");
        }
    }
}
